eslint/artsyPaletteCanonicalImport.js

```diff
@@ -0,0 +1,264 @@
+
+      for (const leafModule of leafModules) {
+        findExports(leafModule)
+      }
+
+      for (const leafType of leafTypes) {
+        findTypeExports(leafType)
+      }
+
+      loaded = true
+    }
+
+    return {
+      ImportDeclaration: function (node) {
+        // starts with @artsy/palette and doesn't match the expected path.
+        const isPaletteImport =
+          node.source &&
+          node.source.type === "Literal" &&
+          node.source.value === "@artsy/palette"
+        if (!isPaletteImport) {
+          return
+        }
+
+        // Match up named imports that come from the same file.
+        const importsMap = new Map()
+        for (const s of node.specifiers) {
+          if (s.type === "ImportSpecifier") {
+            const namedImport = s.local.name
+            const modulePath = exportLookupMap.get(namedImport)
+
+            if (!exportLookupMap.has(namedImport)) {
+              // FIXME: Should the lint rule do anything is there is a
+              // named import that we don't know about?
+              debuglog(`Could not locate export ${namedImport}`)
+            }
+
+            if (!importsMap.has(modulePath)) {
+              importsMap.set(modulePath, [])
+            }
+            importsMap.get(modulePath).push(namedImport)
+          }
+        }
+
+        // Generate new import statements.
+        const imports = []
+        for (const key of importsMap.keys()) {
+          const modulePath = key
+          const nameImports = importsMap.get(key).join(", ")
+          imports.push(`import { ${nameImports} } from "${modulePath}"`)
+        }
+
+        context.report({
+          node,
+          message: `Prefer canonical import form for ${node.source.raw}: ${imports[0]}`,
+          fix: fixer => {
+            return fixer.replaceText(node, imports.join("\n"))
+          },
+        })
+      },
+    }
+  },
+}
+
+// Load the @artsy/palette bundle information.
+function loadPalettePackageJson() {
+  return require(path.join(palettePath, "package.json"))
+}
+
+// Load all the modules that contain the exports we want to validate.
+function locateLeafModules(entryPoint) {
+  const moduleDir = path.dirname(entryPoint)
+  return locateLeafModulesRecursive(moduleDir)
+}
+
+function locateLeafModulesRecursive(moduleDir) {
+  const entries = fs.readdirSync(moduleDir)
+
+  let leafModules = []
+  let leafTypes = []
+  for (const entry of entries) {
+    const entryPath = path.join(moduleDir, entry)
+    const entryStat = fs.statSync(entryPath)
+
+    if (entryStat.isDirectory()) {
+      const results = locateLeafModulesRecursive(entryPath)
+      leafModules = leafModules.concat(results.leafModules)
+      leafTypes = leafTypes.concat(results.leafTypes)
+    }
+
+    // Ignore included files that do not pertain to browser development.
+    if (
+      entryStat.isFile() &&
+      entry.endsWith(".js") &&
+      !entry.endsWith(".ios.js") &&
```

Comment:
This shouldn't be necessary now that https://github.com/artsy/palette/pull/766 has been merged.